editor/src/components/editor/store/collaborative-editing.ts

```diff
@@ -0,0 +1,359 @@
+          // Same value, stop here.
+          return false
+        } else {
+          return applyChanges(fullPath, firstContents, secondContents)
+        }
+      }
+    }
+  }
+  if (isBrowserEnvironment && isFeatureEnabled('Collaboration')) {
+    if (oldContents != newContents) {
+      zipContentsTree(oldContents, newContents, onElement)
+    }
+  }
+
+  return changesToProcess
+}
+
+function applyFileChangeToMap(
+  change: ProjectFileChange,
+  projectContentsMap: CollaborativeEditingSupportSession['projectContents'],
+  mergeDoc: Y.Doc,
+): void {
+  switch (change.type) {
+    case 'DELETE_PATH':
+      const keyIterator = projectContentsMap.keys()
+      let keyIteratorValue = keyIterator.next()
+      while (!keyIteratorValue.done) {
+        const key = keyIteratorValue.value
+        if (
+          key === change.fullPath ||
+          (change.recursive && key.startsWith(`${change.fullPath}/`))
+        ) {
+          projectContentsMap.delete(key)
+        }
+      }
+      break
+    case 'WRITE_PROJECT_FILE':
+      if (
+        change.projectFile.type === 'TEXT_FILE' &&
+        change.projectFile.fileContents.parsed.type === 'PARSE_SUCCESS'
+      ) {
+        const parsedPart = change.projectFile.fileContents.parsed
+        let collabFile: CollabTextFile
+        if (projectContentsMap.has(change.fullPath)) {
+          collabFile = projectContentsMap.get(change.fullPath)!
+          // TODO Handle not text type
+        } else {
+          collabFile = new Y.Map()
+          collabFile.set('type', 'TEXT_FILE')
+          const topLevelElementsArray = new Y.Array<TopLevelElement>()
+          collabFile.set('topLevelElements', topLevelElementsArray)
+        }
+
+        mergeDoc.transact(() => {
+          if (!projectContentsMap.has(change.fullPath)) {
+            projectContentsMap.set(change.fullPath, collabFile)
+          }
+          synchroniseParseSuccessToCollabFile(parsedPart, collabFile)
+        })
+      }
+      break
+    case 'ENSURE_DIRECTORY_EXISTS':
+      // Not necessary to handle right now.
+      break
+    default:
+      assertNever(change)
+  }
+}
+
+export function updateCollaborativeProjectContents(
+  session: CollaborativeEditingSupportSession,
+  projectChanges: ProjectChanges,
+  filesModifiedByElsewhere: Array<string>,
+): void {
+  const projectContentsMap = session.projectContents
+  for (const change of projectChanges.fileChanges.collabProjectChanges) {
+    if (!filesModifiedByElsewhere.includes(change.fullPath)) {
+      applyFileChangeToMap(change, projectContentsMap, session.mergeDoc)
+    }
+  }
+}
+
+export function addHookForProjectChanges(
+  session: CollaborativeEditingSupportSession,
+  dispatch: EditorDispatch,
+): void {
+  session.projectContents.observeDeep((changeEvents) => {
+    // TODO Check that this is the array change before doing anything
+    for (const changeEvent of changeEvents) {
+      switch (changeEvent.path.length) {
+        case 0: {
```

Comment:
would it be a good idea to make these values (0, 1, 2) constants with a more descriptive name?

Reply:
I'm not sure what we'd have for descriptive names as they were kinda identified by just trying various changes and seeing what we got.

Reply:
As an alternative, added an explanatory comment to each case in 0ffcad7b2.